ctsm5.4.036: Complete the FATES-CLM nitrogen coupling

slevis-lmwg opened the pull request:
### Description of changes
For now see the issue #3378 

Corresponding mods on the FATES side: 
https://github.com/NGEET/fates/pull/1472

[Nutrient enabled FATES handbook](https://docs.google.com/document/d/1I35fGDfKTkn9_8Z6qXot7HZf3ICSpLONd-iFEErAh0k/edit?usp=drive_link)
[FATES CLM N coupling](https://docs.google.com/document/d/1mpBtpCLGJpAGw6R3-nGGY92IWRU3ISvaAUKxvcDVJRw/edit?usp=drive_link)

### Specific notes

Contributors other than yourself, if any:
@rgknox @adrifoster @wwieder 

CTSM Issues Fixed (include github issue #):
#3378 

Are answers expected to change (and if so in what way)?
For non-fates tests expect roundoff diffs due to a change in the order of operations in CNNDynamicsMod.F90 as documented [below](https://github.com/ESCOMP/CTSM/pull/3409#issuecomment-4238969834).

Any User Interface Changes (namelist or namelist defaults changes)?
Yes, see changes to CLMBuildNamelist and namelistdefaults.

Does this create a need to change or add documentation? Did you do so?
Yes, no.

- [x] Todo: add a modification to the PRT2 test, make sure that prescribed P uptake is set to 10. This will ensure that there are no P limitations in fates, when FATES becomes coupled to CLM's N cycle (in future PR). i:e: fates_cnp_prescribed_puptake=10

Initial testing performed
...with the first two commits in this PR:
```
PASS ERS_D_Ld30.1x1_brazil.I2000Clm60FatesCrujraRs.derecho_intel.clm-FatesColdPRT2
PASS ERS_D_Ld30.1x1_brazil.I2000Clm60FatesCrujraRs.derecho_intel.clm-FatesCold
```
Later comments point out that these two tests were inadequate at catching problems, and that I switched to two other tests.

slevis-lmwg committed:
Add two 1x1 FatesColdPRT2 tests, one with, one without MIMICS

PASS ERS_D_Ld30.1x1_brazil.I2000Clm60FatesCrujraRs.derecho_intel.clm-FatesColdPRT2
FAIL ERS_D_Ld30.1x1_brazil.I2000Clm60FatesCrujraRs.derecho_intel.clm-FatesColdPRT2--clm-mimicsFatesCold--clm-nofireemis
The latter needs "nofireemis" to work with Fates, but it then dumps core in line 1180 SoilBiogeochemDecompCascadeMIMICSMod.F90

slevis-lmwg committed:
Uncomment if ( .not. use_fates_sp ) section of clmfates_interfaceMod

...calculating these variables:
nf_soil%decomp_npools_sourcesink_col
nf_soil%fates_litter_flux

slevis-lmwg committed:
Replace plant_ndemand(c) * nuptake_prof(c,j) with plant_ndemand_vr(c,j)

slevis-lmwg committed:
New use_fates section for plant_ndemand_vr and plant_ndemand from fates

slevis-lmwg committed:
New use_fates section to set the FATES N uptake fluxes

slevis-lmwg committed:
Change clm_fate from "in" to "inout" argument

slevis-lmwg commented:
With the latest commit, I repeated the two earlier tests and added another to check whether answers have changed from the baseline:
```
PASS ERS_D_Ld30.1x1_brazil.I2000Clm60FatesCrujraRs.derecho_intel.clm-FatesColdPRT2
PASS ERS_D_Ld30.1x1_brazil.I2000Clm60FatesCrujraRs.derecho_intel.clm-FatesCold
FAIL ERP_Ld9.f45_f45_mg37.I2000Clm50FatesRs.derecho_intel.clm-FatesColdAllVars -c /glade/campaign/cgd/tss/ctsm_baselines/ctsm5.3.065
```
The latter fails in case2, after reading the restart file, with a N balance error.

rgknox commented:
Since its failing a FATES configuration that does not have dynamic N, I'm betting that we are running the new uptake portion of the code when we shouldn't.

slevis-lmwg commented on src/soilbiogeochem/SoilBiogeochemCompetitionMod.F90:
```diff
@@ -319,12 +327,51 @@ subroutine SoilBiogeochemCompetition (bounds, num_bgc_soilc, filter_bgc_soilc,nu
+               ! to scale up to column
+               plant_ndemand(c) = 0._r8
+
+               ! We fill the vertically resolved array to simplify some jointly used code
```
Likely still need to resolve N bal. error when parteh == 2

slevis-lmwg replied:
The 3rd test still fails with a N-bal error, and I'm running the 4th test to see what happens:
- I still see indications of badly initialized variables...
- Try uncommenting the harvest lines mentioned in an earlier post
- Consider inconsistent if-statements

slevis-lmwg committed:
Replace if (use_fates) with (is_fates(c)) to add flexibility

slevis-lmwg commented on src/soilbiogeochem/SoilBiogeochemCompetitionMod.F90:
```diff
@@ -991,6 +1038,44 @@ subroutine SoilBiogeochemCompetition (bounds, num_bgc_soilc, filter_bgc_soilc,nu
 
+         ! Set the FATES N uptake fluxes
+
+         if (use_fates) then
```
First two tests listed above still PASS, so I'm resolving this conversation.

slevis-lmwg committed:
Remove two unhelpful 1x1 FatesColdPRT2 tests that I added earlier

slevis-lmwg committed:
New is_fates section for plant_ndemand, now when use_nitrif_denitrif

slevis-lmwg committed:
Add (really, uncomment) FATES npp to enable fixation

slevis-lmwg commented:
Enabled fixation and ran the same tests:
```
PASS ERS_D_Ld30.f45_f45_mg37.I2000Clm50FatesCruRsGs.derecho_intel.clm-FatesColdPRT2 -c /glade/campaign/cgd/tss/ctsm_baselines/fates-sci.1.84.0_api.40.0.0-ctsm5.3.066
PASS ERS_D_Ld30.f45_f45_mg37.I2000Clm50FatesCruRsGs.derecho_intel.clm-FatesColdLUH2 -c /glade/campaign/cgd/tss/ctsm_baselines/fates-sci.1.84.0_api.40.0.0-ctsm5.3.066
```
PRT2 still b4b with the baseline.
LUH2 now DIFF from the baseline.

slevis-lmwg committed:
Uncomment hrv_deadstem_to_prod* lines in clmfates_interfaceMod

...for fates_parteh_mode == 2

slevis-lmwg committed:
Add error check about fates_parteh_mode = 2 and suplnitro = suplnAll

slevis-lmwg committed:
Upd. namelist_defaults for suplnitro when parteh_mode == 1 or /= 1

slevis-lmwg committed:
prt_carbon_allom_hyp -> fates_c_only & prt_cnp_flex_allom_hyp ->fates_cn

These variables originate in fates, so this renaming requires the same
renaming in fates; I will open the corresponding PR very soon

slevis-lmwg committed:
Correction to syntax error caught by gnu and nvhpc, not intel

slevis-lmwg committed:
Update fates paramfile var prescribed_puptake to 1 for PRT2 test

slevis-lmwg commented:
Submitted the same 2 tests
```
./create_test ERS_D_Ld30.f45_f45_mg37.I2000Clm50FatesCruRsGs.derecho_intel.clm-FatesColdLUH2 -c /glade/campaign/cgd/tss/ctsm_baselines/fates-sci.1.84.0_api.40.0.0-ctsm5.3.066
./create_test ERS_D_Ld30.f45_f45_mg37.I2000Clm50FatesCruRsGs.derecho_intel.clm-FatesColdPRT2 -c /glade/campaign/cgd/tss/ctsm_baselines/fates-sci.1.84.0_api.40.0.0-ctsm5.3.066
```
now with the error check suggested in https://github.com/NGEET/fates/issues/1454 (see corresponding commit https://github.com/NGEET/fates/pull/1472/commits/97839f187e44599c3183afa34fd5de16982560a3 in https://github.com/NGEET/fates/pull/1472)

- PRT2 same behavior: variable(s) missing from restart?
- LUH2 FAIL triggered the new error message (turns out it shouldn't have, yet, becase fates_parteh_mode was 1):
```
hlm_name cannot be CLM while prescribed_puptake == 0, because the CLM does not do prognostic phosphorus. Exiting.
ERROR in FatesInterfaceMod.F90 at line 919
```

UPDATE Repeating with 2025/12/10 updates:
- PRT2 same behavior: COMPARE_base_rest
- LUH2 didn't trigger the new error message in its new location and passed, as expected; when I added `fates_parteh_mode = 2`, another similar error got triggered, and Ryan and I decided to keep the latter and remove the former
- Test on izumi 2025/12/16 while derecho is down (gnu, intel, nag):
`ERS_D_Ld30.f45_f45_mg37.I2000Clm50FatesCruRsGs.izumi_*.clm-FatesColdLUH2` FAILs in RUN as per https://github.com/ESCOMP/CTSM/issues/2653, so I tried
`PASS ERS_D_Mmpi-serial_Ld5.1x1_brazil.I2000Clm50FatesRs.izumi_nag.clm-FatesCold` with parteh_mode = 1
`FAIL ERS_D_Mmpi-serial_Ld5.1x1_brazil.I2000Clm50FatesRs.izumi_nag.clm-FatesCold` with parteh_mode = 2 as expected
- [x] Troubleshoot restart issue (SL, but also Ryan offered to look into it)
  - `FAIL ERS_D_Ld30.f45_f45_mg37.I2000Clm50FatesCruRsGs.izumi_nag.clm-FatesColdPRT2` as expected https://github.com/ESCOMP/CTSM/issues/3183 but it may work if I shorten the test to Ld10
  - Return to derecho: UPDATED my fates-cn branches to ctsm5.4.004/sci.1.88.0_api.42.0.0
LUH2 DIFF from latest baseline fates-sci.1.88.0_api.42.0.0-ctsm5.4.001 EXPECTED due to update to .004
PRT2 FAIL COMPARE_base_rest
  - The [first post mentioning the restart problem](https://github.com/ESCOMP/CTSM/pull/3409#issuecomment-3325413045) made me think of rerunning the original PRT2 against the latest baseline `fates-sci.1.89.0_api.43.0.0-ctsm5.4.005` without https://github.com/ESCOMP/CTSM/pull/3409/commits/391a0231f07223a03d0f1f75391253931aefcb03 and that PASSed despite the NLCOMP change of suplnitro from 'ALL' to 'NONE'.
- [x] Add new tests (from meeting with Ryan 2025/12/16):
  - [x] PRT2_synthN (nuptake = 10) to reproduce what used to happen before coupling
  - [x] PRT2_suplnAll that swaps namelist setting suplnitro from NONE to ALL. Initially failed with expected namelist error that we decided to remove:
```
ERROR : CLMBuildNamelist::setup_logic_supplemental_nitrogen() : supplemental Nitrogen (suplnitro) is NOT set to NONE, FATES is on, and FATES-SP is not active, but fates_parteh_mode is 2, so Nitrogen is active; change suplnitro back to NONE
```

slevis-lmwg committed:
Adjust unit conversions to agree with FATES commit b2a1da6

slevis-lmwg committed:
Change fates paramfile var prescribed_puptake to 10 for PRT2 test

slevis-lmwg committed:
Replace fates_cn with fates_cnp throughout

slevis-lmwg committed:
Remove FATES_ERROR_EL from user_nl_clm (out-of-scope fix)

slevis-lmwg committed:
Merge tag 'ctsm5.4.004' into fates-cn

Update crop parameters and maturity requirements

slevis resolved conflicts:
src/fates <-- So far I have not changed .gitmodules to point to my branch
src/soilbiogeochem/SoilBiogeochemCompetitionMod.F90

slevis-lmwg committed:
Merge tag 'ctsm5.4.005' into fates-cn

FATES JSON parameter files

These changes accommodate a JSON format to the FATES parameter file.

Documentation has been updated in the FATES User's Guide:
https://fates-users-guide.readthedocs.io/en/latest/user/Parameter-File.html

fates was updated from sci.1.88.0_api.42.0.0 to sci.1.89.0_api.43.0.0

ESCOMP and NGEET PRs:
https://github.com/ESCOMP/ctsm/pull/3570
https://github.com/NGEET/fates/pull/1493

slevis resolved conflicts:
cime_config/testdefs/testmods_dirs/clm/FatesColdPRT2/shell_commands
cime_config/testdefs/testmods_dirs/clm/FatesColdPRT2/user_nl_clm
src/fates

slevis-lmwg committed:
Add PRT2 test with synthetic N to replicate the pre-CN-coupling setup

slevis-lmwg commented on bld/CLMBuildNamelist.pm:
```diff
@@ -3349,6 +3349,18 @@ sub setup_logic_supplemental_nitrogen {
+    if ( ($parteh_mode == 2) &&  ($suplnitro !~ /NONE/) && not &value_is_true( $nl_flags->{'use_fates_sp'}) ) {
+      $log->fatal_error("supplemental Nitrogen (suplnitro) is NOT set to NONE, FATES is on, " .
+                        "and FATES-SP is not active, but fates_parteh_mode is 2, so Nitrogen is active; " .
+                        "change suplnitro back to NONE");
```
Remove this new check because in parteh_mode = 2 we want to allow either NONE or ALL suplnitro.

slevis-lmwg commented on bld/CLMBuildNamelist.pm:
```diff
@@ -3349,6 +3349,18 @@ sub setup_logic_supplemental_nitrogen {
+    if ( ($parteh_mode == 1) &&  ($suplnitro !~ /ALL/) && not &value_is_true( $nl_flags->{'use_fates_sp'}) ) {
+      $log->fatal_error("supplemental Nitrogen (suplnitro) is NOT set to ALL, FATES is on, " .
+                        "and FATES-SP is not active, but fates_parteh_mode is 1, so Nitrogen is not active. " .
+                        "Change suplnitro back to ALL");
```
Here remove the sp condition because irrelevant.

rgknox committed:
changing namelist type for fates_parteh_mode, part 1

slevis-lmwg committed:
Update two namelist error-checks

slevis-lmwg committed:
Update equivalent to last error-checks, this time in controlMod.F90

slevis-lmwg committed:
Add FatesColdPRT2_suplnAll test and corresponding testmod

rgknox committed:
more name changes for fates CNP and c-only

rgknox committed:
working on passing nl defaults

slevis-lmwg committed:
Merge branch 'fates-cn' into fates-cn

rgknox committed:
fixes to namelist processing of suplnitro

rgknox committed:
more explicit logic on suplnitro with fates

rgknox committed:
more explicit namelist defaults

slevis-lmwg committed:
Remove trailing whitespace

slevis-lmwg committed:
Changes for namelist updates to work

slevis-lmwg committed:
Merge pull request #19 from rgknox/fates-cn

name changes (WIP)

slevis-lmwg commented:
```
./create_test ERS_D_Ld30.f45_f45_mg37.I2000Clm50FatesCruRsGs.derecho_intel.clm-FatesColdPRT2_synthN
./create_test ERS_D_Ld30.f45_f45_mg37.I2000Clm50FatesCruRsGs.derecho_intel.clm-FatesColdLUH2 -c /glade/campaign/cgd/tss/ctsm_baselines/fates-sci.1.89.0_api.43.0.0-ctsm5.4.005
./create_test ERS_D_Ld30.f45_f45_mg37.I2000Clm50FatesCruRsGs.derecho_intel.clm-FatesColdPRT2 -c /glade/campaign/cgd/tss/ctsm_baselines/fates-sci.1.89.0_api.43.0.0-ctsm5.4.005
./create_test ERS_D_Ld30.f45_f45_mg37.I2000Clm50FatesCruRsGs.derecho_intel.clm-FatesColdPRT2_suplnAll
```

slevis-lmwg commented:
@rgknox to follow up...
If this PR and the fates-side PR seem ready to you, then I think we should look for the source of the restart failure. I know you offered to help with that, but let me know if you would like me to start looking into it.

rgknox committed:
Merge tag 'ctsm5.4.019' into fates-cn-merge

Fix years used by generate_gdds

rgknox committed:
merge with  5.3.026

rgknox committed:
Added uptake fluxes to restarts for fates, pass in boundary condition structure

slevis-lmwg committed:
Merge pull request #20 from rgknox/fates-cn

restarting uptake fluxes

slevis-lmwg committed:
Merge branch 'master' into fates-cn

slevis-lmwg commented:
After the latest updates, I started by submitting `SMS.f10_f10_mg37.I2000Clm50BgcCrop.derecho_intel.clm-crop.C.20260406_124851_rch8g6` and this passed. I got pulled in other directions for the day, but my plan is to
- confirm the other tests work...

**FAIL PRT2_synthN RUN**
```
forrtl: severe (408): fort: (2): Subscript #1 of the array PLANT_NH4_UPTAKE_FLUX has value 2 which is greater than the upper bound of 1
cesm.exe           0000000001DFE98F  set_restart_vecto        2713  FatesRestartInterfaceMod.F90
cesm.exe           000000000099C9D5  restart                  1935  clmfates_interfaceMod.F90
cesm.exe           0000000000939B38  clm_instrest              620  clm_instMod.F90
cesm.exe           0000000000D32E1C  restfile_write            132  restFileMod.F90
cesm.exe           0000000000917E69  clm_drv                  1529  clm_driver.F90
cesm.exe           0000000000856DC4  modeladvance              913  lnd_comp_nuopc.F90
```
----
**FAIL LUH2 RUN** turns out to expected as per #3789 
```
negative area           1           1           1  -999.0
ENDRUN: ERROR in EDInitMod.F90 at line 608
```
----
PASS PRT2, just no baseline to compare
PASS PRT2_suplnAll
PASS SMS (as already mentioned above the bullet)
- resolve failures, then start a full aux_clm test

slevis-lmwg commented:
@rgknox two tests failed (PRT2_synthN and LUH2) and three passed. See error msgs in the previous post.

Looking into the first of these. I think we set `allocate(bc_in%plant_nh4_uptake_flux(1,1))` in main/FatesInterfaceMod.F90 because coupled_p_uptake and coupled_n_uptake are false. The failure then occurs in main/FatesRestartInterfaceMod.F90 in this loop:
```
icomp = 0
do_patch: do while(associated(cpatch))
   do while(associated(ccohort))
      if (hlm_parteh_mode == carbon_nitrogen_phosphorus) then
         icomp=icomp+1
         this%rvars(ir_nh4uptakeflux_co)%r81d(io_idx_co) = bc_in(s)%plant_nh4_uptake_flux(icomp,1)
```
My first inclination is to move `icomp = 0` into the first do loop and try the tests again. So far:

FAIL PRT2_synthN RUN
FAIL LUH2 RUN expected
FAIL PRT2 COMPARE_base_rest
FAIL PRT2_suplnAll COMPARE_base_rest
PASS SMS

so my first intuition was wrong, and I will restore FatesRestartInterfaceMod.F90 and try changing the allocate statement in FatesInterfaceMod.F90.

PASS PRT2_synthN RUN
FAIL LUH2 RUN expected
PASS PRT2 COMPARE_base_rest
PASS PRT2_suplnAll COMPARE_base_rest
PASS SMS

Fixed!

slevis-lmwg commented:
Starting aux_clm on derecho + izumi:
`./run_sys_tests -s aux_clm -c ctsm5.4.029 --skip-generate`

RUN failures to investigate (none on izumi) in /glade/derecho/scratch/slevis/tests_0408-165737de
```
ERI_D_Ld9.f45_f45_mg37.I2000Clm60FatesSpCruRsGs.derecho_intel.clm-FatesColdSatPhenCamLndTuningMode
ERP_P128x2_Ld30.f45_f45_mg37.I2000Clm60FatesSpCruRsGs.derecho_intel.clm-FatesColdSatPhen
ERP_P64x2_D_Ld5.f10_f10_mg37.I1850Clm50Bgc.derecho_intel.clm-ciso--clm-matrixcnOn_ignore_warnings
SMS.f45_f45_mg37.I2000Clm60FatesSpRsGs.derecho_nvhpc.clm-FatesColdSatPhen
```
- ERP_P64 reran four times and then a fifth after a clean build and it continues to fail, so I'm abandoning, even though an ERP_D test with the same "matrix" error passed with one rerun.
- ERI, ERP_P128, SMS give `ERROR: fates_parteh_mode=carbon_only must have suplnitro set to suplnAll. ERROR in controlMod.F90 at line 504`. Erik and I looked at CLMBuildNamelist.pm and resolved quickly.
- [x] Erik suggested I add a build-namelist_test.pl test confirming this failure triggers in the build phase.
----
Also, should we expect the DIFFs from baseline in the following tests (some also on izumi, but I did not list them) due to the diff between `sci.1.91.3_api.43.1.0-25-g21c1bcbe` and the baseline `sci.1.91.1_api.43.1.0`? Though I see diffs in non-Fates cases, too. Erik suggests introducing the changes in a b4b step followed by a non-b4b step if possible.
```
    FAIL ERI_D_Ld9.f10_f10_mg37.I1850Clm45Bgc.derecho_gnu.clm-default BASELINE ctsm5.4.029: DIFF
    FAIL ERP_D_P64x2_Ld3.f10_f10_mg37.I2000Clm45BgcCrop.derecho_gnu.clm-no_subgrid_fluxes BASELINE ctsm5.4.029: DIFF
    FAIL ERS_D_Ld20.f45_f45_mg37.I2000Clm50FatesRs.derecho_gnu.clm-FatesColdTwoStreamNoCompFixedBioGeo BASELINE ctsm5.4.029: DIFF
    FAIL ERS_D_Ld6.f10_f10_mg37.I1850Clm45BgcCrop.derecho_gnu.clm-clm50CMIP6frc BASELINE ctsm5.4.029: DIFF
    FAIL LGRAIN2_Ly2_P128x1.f10_f10_mg37.I1850Clm45BgcCrop.derecho_gnu.clm-ciso--clm-cropMonthOutput BASELINE ctsm5.4.029: DIFF
    FAIL LREPRSTRUCT_Ly2_P128x1.f10_f10_mg37.I1850Clm45BgcCrop.derecho_gnu.clm-ciso--clm-cropMonthOutput BASELINE ctsm5.4.029: DIFF
    FAIL SMS_D_Ld5.f10_f10_mg37.I2000Clm50FatesRs.derecho_gnu.clm-FatesCold BASELINE ctsm5.4.029: DIFF
    FAIL SMS_Ld5.f10_f10_mg37.I1850Clm45BgcCrop.derecho_gnu.clm-till--clm-remove_residues BASELINE ctsm5.4.029: DIFF
    FAIL SMS_Ld5_PS.f19_g17.I2000Clm50FatesRs.derecho_gnu.clm-FatesCold BASELINE ctsm5.4.029: DIFF
    FAIL ERI_D_Ld20.f10_f10_mg37.I2000Clm50Fates.derecho_intel.clm-FatesCold BASELINE ctsm5.4.029: DIFF
    FAIL ERI_D_Ld20.f45_f45_mg37.I2000Clm50FatesRs.derecho_intel.clm-FatesColdTwoStream BASELINE ctsm5.4.029: DIFF
    FAIL ERI_D_Ld9.f45_f45_mg37.I2000Clm60Fates.derecho_intel.clm-FatesColdCamLndTuningMode BASELINE ctsm5.4.029: DIFF
    FAIL ERP_D_P64x2_Ld3.f10_f10_mg37.I2000Clm50BgcCru.derecho_intel.clm-noFUN_flexCN BASELINE ctsm5.4.029: DIFF
    FAIL ERP_D_P64x2_Ld3.f10_f10_mg37.I2000Clm50BgcCru.derecho_intel.clm-noFUN_flexCN--clm-matrixcnOn_ignore_warnings BASELINE ctsm5.4.029: DIFF (EXPECTED FAILURE)
    FAIL ERP_Ld9.f45_f45_mg37.I2000Clm50FatesRs.derecho_intel.clm-FatesColdAllVars BASELINE ctsm5.4.029: DIFF
    FAIL ERP_P64x2_D_Ld5.f10_f10_mg37.I1850Clm45BgcCrop.derecho_intel.clm-crop BASELINE ctsm5.4.029: DIFF
    FAIL ERP_P64x2_D_Ld5.f10_f10_mg37.I1850Clm45BgcCru.derecho_intel.clm-ciso BASELINE ctsm5.4.029: DIFF
    FAIL ERP_P64x2_D_Ld5.f10_f10_mg37.I1850Clm45BgcCru.derecho_intel.clm-default BASELINE ctsm5.4.029: DIFF
    FAIL ERP_P64x2_D_Ld5.f10_f10_mg37.IHistClm45BgcCru.derecho_intel.clm-decStart BASELINE ctsm5.4.029: DIFF
    FAIL ERP_P64x2_Ld396.f10_f10_mg37.IHistClm60Bgc.derecho_intel.clm-monthly--clm-matrixcnOn_ignore_warnings BASELINE ctsm5.4.029: DIFF (EXPECTED FAILURE)
    FAIL ERS_D_Ld5.f10_f10_mg37.I2000Clm50Fates.derecho_intel.clm-FatesCold BASELINE ctsm5.4.029: DIFF
    FAIL ERS_D_Ld6.f10_f10_mg37.I1850Clm45BgcCrop.derecho_intel.clm-clm50CMIP6frc BASELINE ctsm5.4.029: DIFF
    FAIL ERS_Ld30.f45_f45_mg37.I2000Clm50FatesRs.derecho_intel.clm-FatesColdFixedBiogeo BASELINE ctsm5.4.029: DIFF
    FAIL ERS_Ld30.f45_f45_mg37.I2000Clm50FatesRs.derecho_intel.clm-FatesColdSizeAgeMort BASELINE ctsm5.4.029: DIFF
    FAIL ERS_Ly5_P128x1.f10_f10_mg37.IHistClm45BgcCrop.derecho_intel.clm-cropMonthOutput BASELINE ctsm5.4.029: DIFF
    FAIL ERS_P128x1_Ld765.f10_f10_mg37.I2000Clm60Fates.derecho_intel.clm-FatesColdNoComp BASELINE ctsm5.4.029: DIFF
    FAIL REP_P64x2_Ld396.f10_f10_mg37.IHistClm60Bgc.derecho_intel.clm-monthly--clm-matrixcnOn_ignore_warnings BASELINE ctsm5.4.029: DIFF (EXPECTED FAILURE)
    FAIL SMS_D_Ld5.f10_f10_mg37.I2000Clm50FatesRs.derecho_intel.clm-FatesCold BASELINE ctsm5.4.029: DIFF
    FAIL SMS_D_Ld65.f10_f10_mg37.I2000Clm45BgcCropQianRs.derecho_intel.clm-FireLi2014Qian BASELINE ctsm5.4.029: DIFF
    FAIL SMS_D_Lm6_P256x1.f45_f45_mg37.I2000Clm50FatesRs.derecho_intel.clm-FatesCold BASELINE ctsm5.4.029: DIFF
    FAIL SMS_Ld5.f10_f10_mg37.I1850Clm45BgcCrop.derecho_intel.clm-crop BASELINE ctsm5.4.029: DIFF
    FAIL SMS_Ld5.f10_f10_mg37.I2000Clm50FatesRs.derecho_intel.clm-FatesCold BASELINE ctsm5.4.029: DIFF
    FAIL SMS_Ln9.f09_f09_mg17.I1850Clm45Bgc.derecho_intel.clm-clm45cam4LndTuningModeZDustSoilErod BASELINE ctsm5.4.029: DIFF
```

slevis-lmwg committed:
Simplify elsif in CLMBuildNamelist.pm to get past failure

slevis-lmwg committed:
Correct minor typo

slevis-lmwg committed:
Add 2 expected failure tests to build-namelist_test.pl

slevis-lmwg committed:
Merge tag 'ctsm5.4.032' into fates-cn

Simplify messaging during docs build

slevis resolved conflicts:
bld/namelist_files/namelist_defaults_ctsm.xml

slevis-lmwg committed:
Update to existing compsets in new PRT2 tests for cases to create

slevis-lmwg commented:
Testing on derecho after the latest updates:
- [x] ./run_sys_tests -s fates -c fates-sci.1.92.0_api.44.0.0-ctsm5.4.031 --skip-generate
- [x] build-namelist_test.pl

Before I try `./run_sys_tests -s aux_clm -c ctsm5.4.032 --skip-generate`, I will look into the diffs from baseline with this test from the above list of failures:
`SMS_Ld5.f10_f10_mg37.I1850Clm45BgcCrop.derecho_intel.clm-crop`
Largest diffs (grepped for "E-0") out of a total of 53 fields with non-zero diffs suggest roundoff diffs:
```
 RMS CH4_SURF_AERE_SAT                2.0582E-10            NORMALIZED  3.9956E-03
 RMS CH4_SURF_DIFF_SAT                8.6639E-14            NORMALIZED  1.6418E-04
 RMS FCH4                             4.7957E-15            NORMALIZED  1.3812E-04
 RMS FCH4TOCO2                        4.7907E-12            NORMALIZED  8.6788E-05
 RMS FCH4_DFSAT                       8.9717E-20            NORMALIZED  2.3733E-07
 RMS NEM                              4.7907E-12            NORMALIZED  1.2779E-04
 RMS SOM_C_LEACHED                    1.3173E-25            NORMALIZED  1.5688E-09
 RMS TOTCOLCH4                        6.1704E-10            NORMALIZED  1.9915E-09
 RMS CONC_O2_SAT                      4.7118E-03            NORMALIZED  4.4369E-02
```
- The test gives similar/same diffs when I reverse the mods in SoilBiogeochemCompetitionMod.F90
- The test returns b4b when I reverse the mods in CNNDynamicsMod.F90
- [x] Starting aux_clm to confirm all DIFFs go away, though Fates tests appear to fail now

slevis-lmwg commented on src/biogeochem/CNNDynamicsMod.F90:
```diff
@@ -239,17 +239,14 @@ subroutine CNNFixation(num_soilc, filter_soilc, &
             else
-               npp = col_lag_npp(c)
+               ! need to put npp in units of gC/m^2/year here first
+               npp = col_lag_npp(c) * secspday * dayspyr
```
This update and line 249 below likely explain (at least in part) the changing answers when running without fates. Keep looking for more potential culprits and then revert all of them (temporarily) to test for b4b.

slevis-lmwg replied:
Good news, the b4b diffs in non-Fates cases go away when I revert the order-of-operation changes in CNNDynamicsMod.F90, and the list of non-Fates failures is the expected:
```
./run_sys_tests -s aux_clm -c ctsm5.4.032 --skip-generate
./cs.status.fails | grep -v NLCOMP | grep -v PASS | grep -v Fates
```
```
    FAIL ERI_Ld41.f10_f10_mg37.I2000Clm60BgcCrop.derecho_gnu.clm-default COMPARE_branch_hybrid (EXPECTED FAILURE)
    FAIL ERP_P64x2_D_Ld3.f10_f10_mt232.IHistClm60BgcCropCrujra.derecho_gnu.clm-default--clm-all_outputs COMPARE_base_rest (EXPECTED FAILURE)
    FAIL ERP_D_P64x2_Ld10.f10_f10_mg37.I2000Clm60Bgc.derecho_intel.clm-Hillslope RUN time=91 (EXPECTED FAILURE)
    PEND ERP_D_P64x2_Ld10.f10_f10_mg37.I2000Clm60Bgc.derecho_intel.clm-Hillslope COMPARE_base_rest
    FAIL ERP_P64x2_Ld396.f10_f10_mg37.IHistClm60Bgc.derecho_intel.clm-monthly--clm-matrixcnOn_ignore_warnings BASELINE ctsm5.4.032: DIFF (EXPECTED FAILURE)
    FAIL ERS_D_Ld7_Mmpi-serial.1x1_smallvilleIA.IHistClm50BgcCropRs.derecho_intel.clm-decStart1851_noinitial RUN time=96 (EXPECTED FAILURE)
    FAIL REP_P64x2_Ld396.f10_f10_mg37.IHistClm60Bgc.derecho_intel.clm-monthly--clm-matrixcnOn_ignore_warnings COMPARE_base_rep2 (EXPECTED FAILURE)
    FAIL REP_P64x2_Ld396.f10_f10_mg37.IHistClm60Bgc.derecho_intel.clm-monthly--clm-matrixcnOn_ignore_warnings BASELINE ctsm5.4.032: DIFF (EXPECTED FAILURE)
    FAIL SMS_D_Ly6_Mmpi-serial.1x1_smallvilleIA.IHistClm45BgcCropQianRs.derecho_intel.clm-cropMonthOutput RUN time=40 (EXPECTED FAILURE)
    FAIL SUBSETDATAPOINT_Ld5_D_Mmpi-serial.CLM_USRDAT.I2000Clm60BgcCropCrujra.derecho_intel.clm-default RUN time=47 (EXPECTED FAILURE)
    FAIL SUBSETDATAREGION_Ld5_D_Mmpi-serial.CLM_USRDAT.I2000Clm60BgcCropCrujra.derecho_intel.clm-default RUN time=66 (EXPECTED FAILURE)
```
(For the record, FATES cases fail with SIGFPE with this code reverted, which is fine because I reverted only to show b4b in all non-Fates cases.)

slevis-lmwg commented on src/soilbiogeochem/SoilBiogeochemCompetitionMod.F90:
```diff
@@ -319,12 +328,53 @@ subroutine SoilBiogeochemCompetition (bounds, num_bgc_soilc, filter_bgc_soilc,nu
-            do fc=1,num_bgc_soilc
-               c = filter_bgc_soilc(fc)      
-               sum_ndemand_vr(c,j) = plant_ndemand(c) * nuptake_prof(c,j) + potential_immob_vr(c,j)
+         bgc_soilc_loop1: do fc = 1, num_bgc_soilc
```
This update likely contributes to the changing answers when running without fates. This one seems more difficult to revert to perform b4b testing, but I will look into it.

slevis-lmwg replied:
Maybe not. Resolving.

slevis-lmwg commented on src/soilbiogeochem/SoilBiogeochemCompetitionMod.F90:
```diff
@@ -337,7 +387,7 @@ subroutine SoilBiogeochemCompetition (bounds, num_bgc_soilc, filter_bgc_soilc,nu
                   fpi_vr(c,j) = 1.0_r8
                   actual_immob_vr(c,j) = potential_immob_vr(c,j)
-                  sminn_to_plant_vr(c,j) = plant_ndemand(c) * nuptake_prof(c,j)
+                  sminn_to_plant_vr(c,j) = plant_ndemand_vr(c,j)
```
This and other changes in this file likely contribute to the changing answers when running without fates...

slevis-lmwg replied:
Probably not, so resolving.

slevis-lmwg commented:
Now repeat `./run_sys_tests -s aux_clm -c ctsm5.4.032 --skip-generate` with the latest commit:
- [x] izumi
- [x] derecho
Three tests ran out of walltime that didn't fail before, so I will not spend additional time on them for now:
```
ERS_Ly5_Mmpi-serial.1x1_smallvilleIA.I1850Clm50BgcCrop.derecho_gnu.clm-ciso_monthly RUN
ERS_Ly5_Mmpi-serial.1x1_smallvilleIA.I1850Clm50BgcCrop.derecho_gnu.clm-ciso_monthly--clm-matrixcnOn RUN
ERS_Ly5_P128x1.f10_f10_mg37.IHistClm60BgcCrop.derecho_intel.clm-cropMonthOutput--clm-matrixcnOn_ignore_warnings RUN
```
Known issue #3840

slevis-lmwg commented:
@rgknox testing seems satisfactory to me, so this PR and its fates counterpart are ready for your final review, unless you think of anything else that we need to work on here.

glemieux commented on cime_config/testdefs/testmods_dirs/clm/FatesColdLUH2/user_nl_clm:
```diff
@@ -27,7 +27,7 @@ hist_fincl1       = 'FATES_NCOHORTS', 'FATES_TRIMMING', 'FATES_AREA_PLANTS',
 'FATES_DEMOTION_CARBONFLUX', 'FATES_PROMOTION_CARBONFLUX',
 'FATES_MORTALITY_CFLUX_CANOPY', 'FATES_MORTALITY_CFLUX_USTORY',
 'FATES_NEP', 'FATES_HET_RESP', 'FATES_FIRE_CLOSS', 'FATES_FIRE_FLUX_EL',
-'FATES_CBALANCE_ERROR', 'FATES_ERROR_EL', 'FATES_LEAF_ALLOC',
```
Do we have an issue to note why this shouldn't be included here?  I could find anything specific outside of a conflict with using it during ST3 or SP mode tests.

slevis-lmwg replied:
I will try putting this back.

slevis-lmwg replied:
Worked!

glemieux commented on cime_config/testdefs/testlist_clm.xml:
```diff
@@ -3900,7 +3900,27 @@
+  <test name="ERS_D_Ld30" grid="f45_f45_mg37" compset="I2000Clm60FatesRs" testmods="clm/FatesColdPRT2_suplnAll">
+    <machines>
+      <machine name="derecho" compiler="intel" category="fates"/>
+      <machine name="lawrencium-lr3" compiler="intel" category="fates"/>
```
@rgknox do you use `lawrencium`?  We should instances of `lawrencium` if we don't think we're going to be using it anytime soon, but that's out of scope for this PR so I'd just make an issue if you agree.

rgknox replied:
yes, we can get rid of lawrencium, I never use that machine

slevis-lmwg replied:
done

glemieux reviewed (approved):
This looks good.  Thanks again to @slevis-lmwg for the walkthrough.  I just have two questions/comments that should be dealt with via other PRs (if at all).

ekluzek commented on bld/namelist_files/namelist_defaults_ctsm.xml:
```diff
@@ -2737,6 +2733,9 @@ lnd/clm2/surfdata_esmf/NEON/ctsm5.4.0/surfdata_1x1_NEON_TOOL_hist_2000_78pfts_c2
 
+<!-- Supplmental Nitrogen mode -->
+<suplnitro >NONE</suplnitro>
+<suplnitro fates_parteh_mode="carbon_only" >ALL</suplnitro>
```
Doesn't this need the addition of:

```suggestion
<suplnitro use_fates=".true." >ALL</suplnitro>
<suplnitro use_fates=".true." fates_parteh_mode="carbon_only" >ALL</suplnitro>
```

ekluzek replied:
OK, it looks like this was an intentional change. I only wonder about making this more obvious in comments though.

slevis-lmwg replied:
@ekluzek I'm happy to add comments. Can you specify what needs clarification? As is, this says that the fates "carbon_only" option sets suplnitro to ALL, while other cases set suplnitro to NONE. Previously, every fates case needed suplnitro set to ALL. Is that what you're looking for? I will add a comment like that preemptively and pls let me know if I'm off the mark.

slevis-lmwg replied:
As testing wraps up, I will resolve remaining conversations in prep. for merging.

ekluzek commented on src/main/clm_varctl.F90:
```diff
@@ -324,9 +324,16 @@ module clm_varctl
-  integer, public            :: fates_parteh_mode = -9                  ! 1 => carbon only
-                                                                        ! 2 => C+N+P (not enabled yet)
-                                                                        ! no others enabled
+  character(len=256), public :: fates_parteh_mode = ''                  ! FATES Plant Allocation Reactions and Transport Hypotheses
```
I really love upgrading this to a character description. So much easier to understand!

ekluzek commented on bld/namelist_files/namelist_definition_ctsm.xml:
```diff
@@ -708,8 +708,8 @@ Toggle to turn on the FATES model
 
-<entry id="fates_parteh_mode" type="integer" category="physics"
-        group="clm_inparm" valid_values="1,2">
+<entry id="fates_parteh_mode" type="char*256" category="physics"
```
You could make the string shorter. If you use SHR_KIND_CS it would be 80 characters which is still longer than needed for this. But, you could use it in the code and have it consistent here.

```suggestion
<entry id="fates_parteh_mode" type="char*256" category="physics"
```

samsrabin replied:
Big big thank you for switching to char instead of int!

slevis-lmwg replied:
Setting to 80 here.

ekluzek commented on src/main/clm_varctl.F90:
```diff
@@ -324,9 +324,16 @@ module clm_varctl
-  integer, public            :: fates_parteh_mode = -9                  ! 1 => carbon only
-                                                                        ! 2 => C+N+P (not enabled yet)
-                                                                        ! no others enabled
+  character(len=256), public :: fates_parteh_mode = ''                  ! FATES Plant Allocation Reactions and Transport Hypotheses
```
I do suggest either hardcoding the length to something shorter (and agree with the value in the namelist defintion above), or use SHR_KIND_CS which has a length of 80.

You can add a use statement to shr_kind, for it, and then use it similar to other string lengths.
```suggestion
  character(len=CS), public :: fates_parteh_mode = ''                  ! FATES Plant Allocation Reactions and Transport Hypotheses
```

``` fortran
use shr_kind_mod, only: CS=>SHR_KIND_CS
```

samsrabin replied:
Yeah, this should use `CS`.

slevis-lmwg replied:
done

ekluzek commented on src/soilbiogeochem/SoilBiogeochemCompetitionMod.F90:
```diff
@@ -545,15 +595,59 @@ subroutine SoilBiogeochemCompetition (bounds, num_bgc_soilc, filter_bgc_soilc,nu
             end do
          end do
 
+         bgc_soilc_loop2: do fc = 1, num_bgc_soilc
```
Add a comment to say what this loop is doing.

samsrabin replied:
Yeah; how does it differ from the `fates1` loop?

slevis-lmwg replied:
The distinction boils down to the if-statement surrounding the two loops:
- bgc_soilc_loop1 is in the `if (.not. use_nitrif_denitrif) then`, while
- bgc_soilc_loop2 in the `else  !----------NITRIF_DENITRIF-------------!`

Given the length of the code in each part, I will add comments clarifying this.

slevis-lmwg replied:
done

ekluzek reviewed (approved):
I did a fairly quick skim, and only have a few suggestions. There was one thing I was concerned about, but realized it was done on purpose. So I only wonder if more comments might help with the supplemental nitrogen setting?

@glemieux had some comments to handle. And we should have him and @samsrabin finish a review as well.

The other thing I wondered about this is how FATES carbon_nitrogen will interact with other parts of Nitrogen in CLM like LUNA? That likely will take some scientific validation, but we might want to have some warnings or comments about some of that?

ekluzek commented on bld/namelist_files/namelist_definition_ctsm.xml:
```diff
@@ -708,8 +708,8 @@ Toggle to turn on the FATES model
+<entry id="fates_parteh_mode" type="char*256" category="physics"
+       group="clm_inparm" valid_values="carbon_only,carbon_nitrogen">
 Switch deciding which nutrient model to use in FATES.
 (Only relevant if FATES is on)
```
The other thing I wonder about is making a note to say that this is experimental.

```suggestion
(Only relevant if FATES is on)
(fates_parteh_mode='carbon_nitrogen' is EXPERIMENTAL and UNSUPPORTED)
```

slevis-lmwg replied:
done

rosiealice commented:
Hi @ekluzek thanks for this. I just wanted to say that this code does not intersect with LUNA, as Vcmax is an input oarameter in FATES and so we don't for now handle any of the acclimation processes that LUNA predicts. This also goes for FUN and the 'Flex-CN' (I'm unsure if we still call it that) functionality.

slevis-lmwg commented:
Note to self: Open new documentation issues to update Technote and User's Guide.

slevis-lmwg commented:
Blocked by https://github.com/NGEET/fates/pull/1472 for now.

slevis-lmwg committed:
Merge tag 'ctsm5.4.034' into fates-cn

bug fix to the FATES land use driver input code

This PR includes a cherry-picked fix from NorESMhub/CTSM#209 to address swap in the
order of rangeland and pasture in the reading of the landuse drivers

slevis resolved conflicts:
src/fates

samsrabin commented on src/main/clm_varpar.F90:
```diff
@@ -53,8 +53,16 @@ module clm_varpar
+  ! are cycled between the two, see corresponding values for fates_parteh_mode:
+  ! bld/namelist_files/namelist_definition_ctsm.xml
+  character(len=256), public, parameter :: clmfates_carbon_only = 'carbon_only'
+  character(len=256), public, parameter :: clmfates_carbon_nitrogen = 'carbon_nitrogen'
```
Again, these should be converted to use length `CS` instead of a magic number.

slevis-lmwg replied:
done

samsrabin reviewed (commented):
Looks great! Just some minor things agreeing with Erik.

slevis-lmwg committed:
Remove passing test from expected failures

slevis-lmwg committed:
Update .gitmodules with the new FATES tag (ie, FATES-side fates-cn)

slevis-lmwg committed:
Remove machine lawrencium from testlist_clm.xml

slevis-lmwg committed:
Add comments about suplnitro in namelist_defaults_ctsm.xml

slevis-lmwg committed:
Change fates_parteh_mode to SHR_KIND_CS (char*80 inst. of 256)

slevis-lmwg committed:
Add comments clarifying the presence of two very similar loops

slevis-lmwg committed:
Merge remote-tracking branch 'escomp/master' into fates-cn

slevis-lmwg commented:
- [x] `./run_sys_tests -s aux_clm -c ctsm5.4.035 -g ctsm5.4.036`
- [x] `./run_sys_tests -s fates -c fates-sci.1.92.4_api.45.0.0-ctsm5.4.036 --skip-generate`
- [x] build-namelist_test
- [x] ChangeLog

slevis-lmwg committed:
Update total number of tests in build-namelist_test.pl

slevis-lmwg committed:
Draft updates to ChangeLog/ChangeSum

slevis-lmwg committed:
Revert 25ea275 by putting FATES_ERROR_EL back in user_nl_clm
